torch/csrc/api/include/torch/data/datasets/chunk.h

```diff
@@ -0,0 +1,304 @@
+#pragma once
+
+#include <c10/util/Exception.h>
+#include <torch/data/datasets/base.h>
+#include <torch/data/example.h>
+#include <torch/data/samplers/base.h>
+#include <torch/data/samplers/sequential.h>
+
+#include <algorithm>
+#include <mutex>
+#include <optional>
+#include <thread>
+
+/// A dataset that supports loading an entire chunk of data.
+///
+/// A chunk could be an entire file, such as an audio data file or an image,
+/// or part of a file in the case of a large text file split based on seek
+/// positions. ChunkDataSet extends the DataSet functionality to read an
+/// antire chunk at once.
+
+namespace torch {
+namespace data {
+namespace datasets {
+
+/// A dataset that supports loading an entire chunk of data.
+///
+/// A chunk could be an entire file, such as an audio data file or an image,
+/// or part of a file in the case of a large text file split based on seek
+/// positions. ChunkDataSet extends the DataSet functionality to read an
+/// antire chunk at once.
+template <
+    typename Self,
+    typename Batch = std::vector<Example<>>,
+    typename ChunkSampler = samplers::SequentialSampler,
+    typename ExampleSampler = samplers::SequentialSampler>
+class ChunkDataSet : public BatchDataset<Self, Batch> {
+ public:
+  /// Read an entire chunk. A derived class needs to override this method.
+  virtual Batch read_chunk(size_t chunk_index) = 0;
+
+  /// Return the chunk count. A derived class needs to override this method.
+  virtual size_t get_chunk_count() = 0;
+
+  void initialize() {
+    /// After the data set is created, we need to call this method before
+    /// getting batch data. This method will use some infomation that is not yet
+    /// available in the constructor to configure chunk sampler and start
+    /// working thread.
```

Comment:
We need to get this method as a standalone PR. Hopefully, this is the only change we need for the dataset API.